### PR TITLE
LSA knob to MAD-X script converter

### PR DESCRIPTION
**Pull request by fsoubelet**
Adds a script which can take an `LSA` knob and an optics it is defined for to create both a definition TFS file and a `MAD-X` script which will reproduce the knob in simulations.

The script can also be fed a text file with many knobs definitions as well as their trim values, and run for all of these knobs. See examples in the module docstring.

### CODE REVIEW

File: pylhc/machine_settings_info.py [hunk 1]
```diff
@@ -77,38 +80,22 @@ def _get_params() -> dict:
...
-            action="store_true",
-            help="Write summary into log (automatically done if no output path is given)."),
+            help=("UTC Time as 'Y-m-d H:M:S.f' format or AccDatetime object." " Defines the beginning of the time-range."),
+        ),
+        knobs=dict(default=None, nargs="+", type=str, help="List of knobnames."),
+        accel=dict(default="lhc", type=str, help="Accelerator name."),
+        output_dir=dict(default=None, type=PathOrStr, help="Output directory."),
+        knob_definitions=dict(action="store_true", help="Set to extract knob definitions."),
+        source=dict(type=str, default="nxcals", help="Source to extract data from."),
+        log=dict(action="store_true", help="Write summary into log (automatically done if no output path is given)."),
```
Comment (JoschD) [inline pylhc/machine_settings_info.py:98]: I hate your black. At least check what you are changing... (every time, man)

File: pylhc/machine_settings_info.py [hunk 2]
```diff
@@ -251,10 +242,7 @@ def log_summary(acc_time: AccDatetime, bp_info: DotDict,
         f"  Descr.:     {bp_info.Description}\n"
     )
     if optics_info is not None:
-        summary += (
-            f"Optics:       {optics_info.Name}\n"
-            f"  Start:      {optics_info.StartTime.utc_string[:-7]} UTC\n"
-        )
+        summary += f"Optics:       {optics_info.Name}\n" f"  Start:      {optics_info.StartTime.utc_string[:-7]} UTC\n"
```
Comment (JoschD) [inline pylhc/machine_settings_info.py:245]: every goddamn time

File: pylhc/machine_settings_info.py [hunk 3]
```diff
@@ -283,21 +275,27 @@ def write_summary(
...
-    info_tfs.headers = OrderedDict([
-        ("Hint:",                           "All times given in UTC."),
-        (const.head_accel,                  accel),
-        (const.head_time,                   acc_time.cern_utc_string()),
-        (const.head_beamprocess,            bp_info.Name),
-        (const.head_fill,                   bp_info.Fill),
-        (const.head_beamprocess_start,      bp_info.StartTime.cern_utc_string()),
-        (const.head_context_category,       bp_info.ContextCategory),
-        (const.head_beamprcess_description, bp_info.Description),
-        ])
```
Comment (JoschD) [inline pylhc/machine_settings_info.py:295]: every ...

File: pylhc/lsa_to_madx.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        help="Name of a file with knob names and strength factors to use. "
+        "A single knob should be written per line, and lines starting with a # "
+        "character will be ignored.",
+    )
+    return parser.parse_args()
+
+
+if __name__ == "__main__":
+    options = _get_args()
+    lsa_optics = options.optics
```
Comment (JoschD) [inline pylhc/lsa_to_madx.py:196]: you don't want to put the code below in a `main()`?
Reasons?

**Review by JoschD — CHANGES_REQUESTED** [review]

**Comment by JoschD** [comment]
why is your diff coverage so low, with all of these tests? xD

File: .gitignore [hunk 5]
```diff
@@ -1,7 +1,261 @@
...
+
+# Neovim
+.nvimlog
+
+/.idea/codeStyles/codeStyleConfig.xml
+/.idea/misc.xml
+/.idea/modules.xml
+/.idea/inspectionProfiles/profiles_settings.xml
+/.idea/vcs.xml
+/.idea/PyhDToolkit.iml
```
Comment (mihofer) [inline .gitignore:250]: I wonder where this gitignore came from ...

Comment (fsoubelet) [reply .gitignore:250]: Ah True, I could clean this

File: pylhc/lsa_to_madx.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+    """
+    knob_lines = [line for line in Path(knobs_file).read_text().splitlines() if not line.startswith("#")]
+    results = {}
+
+    for line in knob_lines:
+        knob = line.split()[0]  # this is the knob name
+        try:  # catch the trim value, which is a number after the knob name
+            trim = float(line.split()[1])
+            results[knob] = trim
+        except IndexError:  # the is no number, we default the trim value to 1.0
```
Comment (mihofer) [inline pylhc/lsa_to_madx.py:121]: the trim value is not a number?

File: pylhc/lsa_to_madx.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        trim (float): The trim value to write for the knob. Defaults to 1.
+
+    Returns:
+        The complete ``MAD-X`` script to reproduce the knob, as a string.
+    """
+    LOG.debug(f"Determining MAD-X commands to reproduce knob '{lsa_knob}'")
+    change_commands = [f"! Start of change commands for knob: {lsa_knob}"]
+
+    # Set this to 1 by default but can be changed by the user to reproduce a given trim
+    knob_itself = lsa_knob.split("/")[1]  # without the LHCBEAM/ part
```
Comment (mihofer) [inline pylhc/lsa_to_madx.py:146]: if this is convention, maybe it's worth putting it as some sort of global regex str at the top

File: pylhc/lsa_to_madx.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    change_commands.append(f"! End of change commands for knob: {lsa_knob}\n")
+    return "\n".join(change_commands)
+
+
+# ----- Script Part ----- #
+
+
+def _get_args():
+    """Parse Commandline Arguments."""
+    parser = argparse.ArgumentParser(
```
Comment (mihofer) [inline pylhc/lsa_to_madx.py:164]: why argparse and generic parser? with generic parser you could make it a fct and combine extract knobs and this

File: tests/inputs/lsa_to_madx/LHCBEAM_MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_knob.madx [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+kq8.l1b1     =    kq8.l1b1     +     4.70197155664e-05     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
+kq10.r1b1    =    kq10.r1b1    +    -1.96085402422e-05     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
+kq4.l1b1     =    kq4.l1b1     +     7.35134890419e-05     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
+kq4.r1b1     =    kq4.r1b1     +     4.70408231195e-05     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
+kq10.l1b1    =    kq10.l1b1    +    -2.47050375037e-05     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
+kq7.r1b1     =    kq7.r1b1     +     -2.3806014724e-05     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
+kq9.l1b1     =    kq9.l1b1     +     0.000118330906844     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
+kq5.l1b1     =    kq5.l1b1     +    -0.000214221465285     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
+kq6.r1b1     =    kq6.r1b1     +     0.000218127839616     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
+kq6.l1b1     =    kq6.l1b1     +     0.000125226986711     * MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_trim;
```
Comment (mihofer) [inline tests/inputs/lsa_to_madx/LHCBEAM_MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_knob.madx:19]: for MAD-X, it is safer to use + ( VARIABLE of unknown sign  ) .. if you have 1 + -1, the parser not like it

File: pylhc/lsa_to_madx.py (hunk 6 above)
Comment (fsoubelet) [reply pylhc/lsa_to_madx.py:121]: It is, but the script accepts no value provided and will default to 1

File: tests/inputs/lsa_to_madx/LHCBEAM_MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_knob.madx (hunk 9 above)
Comment (fsoubelet) [reply tests/inputs/lsa_to_madx/LHCBEAM_MD_ATS_2022_05_04_B1_RigidWaitsShift_IP1pos_knob.madx:19]: Good to know, this was indeed mentioned in the meeting this morning. I will make that change.

File: pylhc/lsa_to_madx.py (hunk 7 above)
Comment (fsoubelet) [reply pylhc/lsa_to_madx.py:146]: Regex :(
As far as I know it is an almost convention, sometimes it is LHCBEAM1 or LHCBEAM2. I will see for a regex

File: pylhc/lsa_to_madx.py (hunk 8 above)
Comment (fsoubelet) [reply pylhc/lsa_to_madx.py:165]: I'll keep this script simple for now, if we want to do this in the future we can always change the parser without breaking functionality

**Review by mihofer — APPROVED** [review]

File: pylhc/lsa_to_madx.py (hunk 7 above)
Comment (JoschD) [reply pylhc/lsa_to_madx.py:146]: honestly, I would go with `lsa_knob.split("/")[-1]` . In that case it doesn't matter if there is a `/` in the name or not (only gets problematic when there is two, but that is even less likely.)

**Pull request merged**